### PR TITLE
Add support for optional attendees for a meeting

### PR DESCRIPTION
**Pull request by ak47na**
Add functionality of optional attendees:
If one or more time slots exists s.t. both mandatory and optional attendees can attend, the query function returns those time slots. 
Otherwise, it returns the time slots that fit just the mandatory attendees.

Add `Collection<String> optionalAttendees` parameter to the query function in current branch in FindMeetingQuery.java
Rename the `query` function from branch **meetingAlgorithm** to `queryWithoutOptionalAttendees`

First, the `query` function searches the time slots for a new `MeetingRequest` object that considers both optional and mandatory attendees as mandatory:
`result = queryWithoutOptionalAttendees(events, newRequest);`
If `result` is empty, `query` will return the time slots that fit just the mandatory attendees:
`queryWithoutOptionalAttendees(events, request);`

Note: this branch will be merged after **meetingAlgorithm** is merged

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java [hunk 1]
```diff
@@ -93,4 +95,17 @@ private static boolean equals(Event a, Event b) {
     // interface documentation, equals will check for set-equality across all set implementations.
     return a.title.equals(b.title) && a.when.equals(b.when) && a.attendees.equals(b.attendees);
   }
+  /**
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java:99]: Nit: missing new line
```suggestion
  
  /**
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java [hunk 2]
```diff
@@ -93,4 +95,17 @@ private static boolean equals(Event a, Event b) {
...
+   */
+  public boolean containsRequestedAttendees(Collection<String> requestedAttendees) {
+    Optional result = 
+    requestedAttendees.parallelStream()
+                      .filter(requestedAttendee -> attendees.contains(requestedAttendee) == true)
+                      .findAny();
+    if (result.isPresent()) {
+      return true;
+    }
+    return false;
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Could you just do `return result.isPresent()`?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,119 @@
...
+    }
+  }
+
+  private void precomputePrefixSum(ArrayList<Integer> meetings) {
+    // compute the sum starting from the second element because the prefix sum for the first 
+    // element is the first element 
+    for (int i = TimeRange.START_OF_DAY + 1; i <= TimeRange.END_OF_DAY; ++ i) {
+      meetings.set(i, meetings.get(i - 1) + meetings.get(i));
+    }
+  }
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Optional: Something which might make these easier to test is to consider converting these to pure functions without any side effects.  This will make them idempotent - i.e. you can call them multiple times and you get the same result each time.

At the moment for these functions they have the "side effect" of each time you call them they change value of `meetings`.  So what happens if you call it say 2 or 3 times in a row and each time they try to update `meetings` again?  Could bad things happen?  Maybe, or maybe not - it is hard to be sure without reading the entire class to see where this function is used, and what data is passed into it.

So to make these functions pure (no "side effects"/idempotent), instead of changing the member `meetings` directly in these functions, they could return a new `ArrayList<Integer>` which is then used by the returning code to update the variable, e.g.

```java
public Collection<TimeRange> queryWithoutOptionalAttendees(Collection<Event> events, MeetingRequest request) {
  // ...
  meetings = precomputePrefixSum(meetings);
  // ...
}

private ArrayList<Integer> precomputePrefixSum(ArrayList<Integer> meetings) {
  // Instead of directly changing member variable meetings, create a copy, modify that, then return the modified copy
}

```

By making these functions idempotent like this, you can call them over and over and they wont modify the internal state of the class (unless you deliberately use the return value to update the internal state with an assignment like in my example above) - this makes them easier to think about because you do not need to know what else is happening to `meetings` elsewhere in the class - you only need to care about what comes in to the function as arguments, and what it returns.

As well as making the code easier to read and reason about, this also makes them much easier to test since you do not need to set up the entire class's state each time either because you know the function only cares about its own inputs and outputs.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,119 @@
...
+                            .collect(Collectors.toList())
+                            .forEach(eventWhen -> updateNumberOfMeetings(meetings, eventWhen));
+
+    // after all events are processed, for each minute x, meetings[x] will represent the number of
+    // meetings that start at minute x, minus the number of meetings that end right before minute x
+
+    // precompute the prefix sum on the meetings array s.t meetings[x] will represent the number of
+    // meetings that happen at minute x:
+    // (before computing the sum):  [ .... +1 ........ -1 .... ]
+    // (after computing the sum) :  [..... +1 +1 ... +1 0 .... ]
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This level of commenting is great - good work. :+1: 

These comments are explaining at a high level what is going on without someone having to "run the code in their heads", with some examples. This is exactly the sort of thing we like.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,119 @@
...
+    // mark that a new meeting starts at when.start()
+    meetings.set(when.start(), meetings.get(when.start()) + 1); 
+    // mark that a new meeting ends right before when.end() 
+    if (when.end() <= TimeRange.END_OF_DAY) {
+      // only update the end time of events that end before the day ends
+      meetings.set(when.end(), meetings.get(when.end()) - 1);
+    }
+  }
+
+  private void precomputePrefixSum(ArrayList<Integer> meetings) {
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please add a comment for this function.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -265,10 +355,52 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request, NO_ATTENDEES);
     Collection<TimeRange> expected = Arrays.asList();
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void noMandatoryAttendees() {
+    // No mandatory attendees, just two optional attendees with one gap in their schedules.
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:368]: Good - it is always sensible to test the "boundary cases" of our business logic.

So classic boundary cases are things like -1,0,1, MAX_INTEGER etc - i.e. you are testing  all the places where the logic might fall apart because you are testing outside the limits of what you expect to get, rather than just the "happy day" scenario where you just test with "normal" data.

Perhaps a couple of other boundary conditions that I am not sure are covered already could be:
 - what if there are no events?
 - what if the day is only a half-day? or zero length? (e.g. weekend day?)
 - what if the events last longer than the day?

Sometimes it feels like a waste of time to test these boundary cases because "this can never happen" ... but of course we know that with code sometimes things we think cannot happen actually do :-)  Especially in the future when we come back to make a change or fix a bug and we've forgotten what we were doing - those boundary checks in our tests might save us from making a big mistake.

**Review by ghost — APPROVED** [review]
Looks good - good work on updating the commenting and glad to see the streams API in use (did you find it helps with code readability?  Easier to program?)

One or two things for you to have a think about - don't worry about getting them in for this PR, but perhaps for future ones.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (ak47na) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thank you for the idea.
 I changed the code to 
```
public Collection<TimeRange> queryWithoutOptionalAttendees(Collection<Event> events, MeetingRequest request) {
  // ...
  return findAvailableTimeRanges(precomputePrefixSum(meetings), (int)request.getDuration());
}
```
Using `meetings = precomputePrefixSum(meetings);` will result in a compilation error (meetings is used in a lambda function and if I understood it correctly, it must be final or effectively final).

Is there a better way to change the code to avoid the error?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (ak47na) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I added a comment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java (hunk 2 above)
Comment (ak47na) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Yes, I've changed it.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (ak47na) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:368]: Thank you for the tip. I added a test with no events and two that have events lasting more than the day. 

I'm not sure if the case when the day is only a half-day or zero length would be valid, but I will think about it.
